torch/csrc/jit/codegen/cuda/kernel.cpp

```diff
@@ -28,81 +30,125 @@ static int ceilDiv(const int a, const int b) {
       auto o_dim = val.dim();
       TORCH_CHECK(b_dim >= o_dim);
-      int64_t* sizes = reinterpret_cast<int64_t*>(buffer_ptr);
-      int64_t* strides = &sizes[b_dim];
       for (int i = 0; i < b_dim; i++) {
-        sizes[i] = broadcasted_size->at(i);
+        sizes.push_back(broadcasted_size->at(i));
         int index = i + o_dim - b_dim;
         if (index < 0) {
-          strides[i] = 0;
+          strides.push_back(0);
         } else if (val.sizes()[index] == sizes[i]) {
-          strides[i] = val.strides()[index];
+          strides.push_back(val.strides()[index]);
         } else {
           TORCH_CHECK(
               val.sizes()[index] == 1,
               "Not compatible dimension size for broadcast");
-          strides[i] = 0;
+          strides.push_back(0);
         }
       }
-      buffer_ptr = &strides[b_dim];
     } else {
       auto o_dim = val.dim();
-      int64_t* sizes = reinterpret_cast<int64_t*>(buffer_ptr);
-      int64_t* strides = &sizes[o_dim];
       for (decltype(val.dim()) i{0}; i < o_dim; i++) {
-        sizes[i] = val.sizes()[i];
-        strides[i] = val.strides()[i];
+        sizes.push_back(val.sizes()[i]);
+        strides.push_back(val.strides()[i]);
       }
-      buffer_ptr = &strides[o_dim];
     }
   }
+};
+
+struct KernelArgumentHolder {
+private:
+  std::vector<ArgAbstract*> arguments;
+  std::vector<char> buffer;
+  std::vector<void*> arg_ptrs;
+  bool changed = true;
+public:
+  virtual ~KernelArgumentHolder(){
+    for(auto  arg: arguments)
+      delete arg;
+  }
+
+  // Push a tensor to the arguments
+  void push(
+      const at::Tensor& val,
+      c10::optional<at::IntArrayRef> broadcasted_size = c10::nullopt) {
+    changed = true;
+    ExtractSizeStride ess (val, broadcasted_size);
+    int nDims = ess.sizes.size();
+    
+    c10::ScalarType dtype = val.scalar_type();
+    TensorArgAbstract *tensor_arg = getTensorArg(dtype, nDims);
+    tensor_arg->setPointer(val.data_ptr());
+    for(int i=0; i<nDims; i++){
+      tensor_arg->setSize(i, ess.sizes[i]);
+      tensor_arg->setStride(i, ess.strides[i]);
+    }
+    arguments.push_back(tensor_arg);
+  }
+
+  // Push a scalar or integer to the arguments
+  void push(const IValue& val) {
+    changed = true;
+    TORCH_INTERNAL_ASSERT(val.isScalar(),
+      "Tried to push an arg to run in a fused kernel, expected a scalar but got, ", val);
+    switch(val.toScalar().type()){
+      case(c10::ScalarType::Double):
+        arguments.push_back(new FloatArg((float) val.toDouble()));
+        return;
+      case(c10::ScalarType::Long):
+        arguments.push_back(new IntArg((int) val.toInt()));
+        return;
+      default:
+        TORCH_INTERNAL_ASSERT(false,
+        " Tried to create argument to send to a fused kernel, but got an expected type: ");
+    }
+          TORCH_INTERNAL_ASSERT(false,
+      " Tried to create argument to send to a fused kernel, but got a non-tensor, non-scalar type.");
```

Comment:
It's fine if we decide to assert on `val.isScalar()` instead of pass forward it to the other overload without broadcast shape. But the comment here shouldn't claim that `val` is a `non-tensor`